ddtrace/tracer/handler.go

```diff
@@ -0,0 +1,235 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+package tracer
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+)
+
+type traceWriter interface {
+	// add adds a trace to the current payload being constructed by the handler.
+	add([]*span)
+
+	// flush causes the handler to send its current payload. The flush can happen asynchronously, but
+	// the handler must be ready to accept new traces with add when flush returns.
+	flush()
+
+	// stop shuts down the traceWriter, ensuring all payloads are flushed.
+	stop()
+}
+
+type agentTraceWriter struct {
+	config *config
+
+	payload *payload
+
+	// climit limits the number of concurrent outgoing connections
+	climit chan struct{}
+
+	// wg waits for all goroutines to exit when stopping.
+	wg sync.WaitGroup
+
+	// prioritySampling refers to the tracer's priority sampler.
+	prioritySampling *prioritySampler
+}
+
+var _ traceWriter = &agentTraceWriter{}
+
+func newAgentTraceWriter(c *config, s *prioritySampler) *agentTraceWriter {
+	return &agentTraceWriter{
+		config:           c,
+		payload:          newPayload(),
+		climit:           make(chan struct{}, concurrentConnectionLimit),
+		prioritySampling: s,
+	}
+}
+
+func (h *agentTraceWriter) add(trace []*span) {
```

Comment:
Is the root guaranteed to always be at index 0? If yes, we could apply the priority sampling here and remove it from the `tracer` struct completely.

Reply:
Actually, after playing around with this, I realized that we can't make changes (such as adding a priority) to a `span` after it is finished, and spans get finished before they are added to the payload.

We could make this change, but I think it would be larger and more extensive than I initially thought.

Reply:
Good point. In that case: sorry to derail the work. Resolving.